runtime/commonMain/src/kotlinx/serialization/json/internal/JsonReader.kt

```diff
@@ -207,16 +225,14 @@ internal class JsonReader(private val source: String) {
         length = 0 // in buffer
         var currentPosition = startPosition + 1
         var lastPosition = currentPosition
-        val length = source.length
         while (source[currentPosition] != STRING) {
-            if (currentPosition >= length) fail("Unexpected EOF", currentPosition)
             if (source[currentPosition] == STRING_ESC) {
                 appendRange(source, lastPosition, currentPosition)
                 val newPosition = appendEsc(source, currentPosition + 1)
                 currentPosition = newPosition
                 lastPosition = newPosition
-            } else {
-                currentPosition++
+            } else if (++currentPosition >= source.length) {
+                fail("EOF", currentPosition)
```

Comment:
```suggestion
                fail("Unexpected EOF", currentPosition)
```

Reply:
This change is deliberate, the resulting text will be: `kotlinx.serialization.json.JsonDecodingException: Unexpected JSON token at offset 9: EOF`